src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -289,8 +289,22 @@ public MethodIL GetMethodILWithInlinedSubstitutions(MethodIL method)
                         if (delta >= 0 && delta < ehRegion.TryLength)
                         {
                             if (ehRegion.Kind == ILExceptionRegionKind.Filter)
+                            {
                                 offsetsToVisit.Push(ehRegion.FilterOffset);
 
+                                // Filter must end with endfilter, so ensure we don't accidentally remove it.
+                                // ECMA-335 dictates that filter block starts at FilterOffset and ends at HandlerOffset.
+                                int expectedEndfilterLocation = ehRegion.HandlerOffset - 2;
+                                bool isValidFilter = expectedEndfilterLocation >= 0
+                                    && (flags[expectedEndfilterLocation] & OpcodeFlags.InstructionStart) != 0
+                                    && methodBytes[expectedEndfilterLocation] == (byte)ILOpcode.prefix1
+                                    && methodBytes[expectedEndfilterLocation + 1] == unchecked((byte)ILOpcode.endfilter);
+                                if (isValidFilter)
+                                {
+                                    flags[expectedEndfilterLocation] |= OpcodeFlags.VisibleBasicBlockStart | OpcodeFlags.Mark;
```

Comment:
Does this mean we will keep the endfilter always (and thus the filter block), or does this preserve the ability to remove the entire filter block?

Reply:
This code runs when we're looking at EH regions associated with a marked instruction. So if the block covered by the filter got trimmed, the filter will be trimmed too.

Reply:
I finally wrapped my head around how this works in the linker, but the implementation in NativeAOT is different... I'll have to spend some time trying to understand it as well.

Reply:
The general approach is the same as in ILLinker - seek backwards in the instruction stream and try to come up with a constant value. It doesn't try to reformat the IL after the fact (just nop things out) because RyuJIT is very efficient at getting rid of nops. We don't have to adjust EH regions or recompute branch offsets as a result.